doc/port_auto_neg/auto-fec.md

```diff
@@ -0,0 +1,202 @@
+```
+
+### Configuration and management
+
+#### Config command
+
+No new CLI commands are introduced to support auto FEC. The existing 'config interface fec' is extended to have additional option called 'auto'. If this option is set without autoneg enabled, it will default to 'none'
+
+```
+Format:
+  config interface fec <interface_name> <mode>
+
+Arguments:
+  interface_name: name of the interface to be configured. e.g: Ethernet0
+  mode: none/rs/fc/auto
+
+Example:
+  config interface fec Ethernet0 auto
+
+Return:
+  error message if interface_name or mode is invalid otherwise empty
+```
+
+If a vendor doesn't support auto mode, an error will be thrown when 'auto' mode is configured.
+```
+sonic# config interface fec Ethernet0 auto
+fec auto is not in ['rs', 'fc', 'none']
+```
+
+#### Show command
+
+Currently the configured FEC mode is displayed in "show interfaces status" command. With the introduction of mode 'auto' it becomes necessary to display operational FEC. For modes other than auto, the config will match the operational FEC. In case of mode 'auto' the operational FEC will be queried from SAI during oper up sequence using SAI_PORT_ATTR_OPER_FEC. This will be updated to state_db PORT_TABLE field "fec". The value will contain the operational FEC and have the config auto in paranthesis as shown below.
+The show command will query the state DB and if there is FEC field, it would be displayed.
+If the field is not present, the config will be displayed, similar to the legacy behavior.
+
+````
+show interfaces  status
+   Interface            Lanes    Speed    MTU    FEC       Alias          Vlan    Oper    Admin             Type    Asym PFC
+------------  ---------------  -------  -----  ---------   -------  ------------  ------  -------  ---------------  ----------
+   Ethernet0          0,1,2,3     100G   9100    N/A       etp1        routed      up       up   QSFP+ or later         N/A
+   Ethernet4          4,5,6,7     100G   9100     rs       etp2  PortChannel1      up       up   QSFP+ or later         N/A
+   Ethernet8        8,9,10,11     100G   9100   rs(auto)   etp3        routed      up       up   QSFP+ or later         N/A
+````
+
+### YANG model changes
+
+The yang model will be modified to accept additional value 'auto' for FEC
+
+```
+leaf fec {
+	type string {
+		pattern "rs|fc|none|auto";
+	}
+}
+```
+
+### Warmboot and Fastboot Considerations
+
+No impact to warm or fastboot
+
+### Restrictions/Limitations
+
+Since previously the behavior of FEC with autoneg is undefined, if a vendor always gives precedence for auto negotiated FEC over user configured FEC, implementing the new attribute may lead to backward incompatibility issues when SAI_PORT_ATTR_AUTO_NEG_FEC_MODE_OVERRIDE is implemented. Since orchagent will always set SAI_PORT_ATTR_AUTO_NEG_FEC_MODE_OVERRIDE to true when FEC is explicitly configured and the attribute is supported by SAI, there will be difference in behavior before and after upgrade.
```

Comment:
`SAI_PORT_ATTR_AUTO_NEG_FEC_MODE_OVERRIDE` support by the platform itself means that the platform is going to OVERRIDE its existing behavior. isn't it? If so, the difference in behavior is expected...i mean it will be a breaking change though...If the platform choose to keep the original behavior, then it means the platform is not supporting the override SAI attribute in which case why support override attribute?

Reply:
If there is  difference in behavior, it should be also made backward compatible through db_migrator. It is upto the vendors to decide whether they want to support new attribute and introduce db_migrator if they need it for solving the backward incompatibility or avoid introducing the attribute altogether.